docs/proposals/runtime-client-server.md

```diff
@@ -0,0 +1,202 @@
+  // all containers in the pod. Clients of this interface assume the containers
+  // statuses in a pod always have a deterministic ordering (eg: sorted by name).
+  GetPodStatus(pod *PodMeta) (*api.PodStatus, error)
+
+  // PullImage pulls an image from the network to local storage using the supplied
+  // secrets if necessary.
+  PullImage(image ImageSpec, pullSecrets []api.Secret) error
+  // IsImagePresent checks whether the container image is already in the local storage.
+  IsImagePresent(image ImageSpec) (bool, error)
+  // ListImages gets all images currently on the machine.
+  ListImages() ([]Image, error)
+  // Removes the specified image.
+  RemoveImage(image ImageSpec) error
+
+  // GetContainerLogs returns logs of a specific container. By
+  // default, it returns a snapshot of the container log. Set 'follow' to true to
+  // stream the log. Set 'follow' to false and specify the number of lines (e.g.
+  // "100" or "all") to tail the log.
+  GetContainerLogs(pod *PodMeta, containerID ContainerID, logOptions *api.PodLogOptions, stdout, stderr io.Writer) (err error)
+  // AttachContainer attachs to the container
+  AttachContainer(id ContainerID, stdin io.Reader, stdout, stderr io.WriteCloser, tty bool) (err error)
+  // ExecInContainer runs the command in the container of the specified pod
+  // Attaches the processes stdin, stdout, and stderr. Optionally uses a
+  // tty.
+  ExecInContainer(containerID ContainerID, cmd []string, stdin io.Reader, stdout, stderr io.WriteCloser, tty bool) error
+  // Forward the specified port from the specified pod to the stream.
+  PortForward(pod *Pod, port uint16, stream io.ReadWriteCloser) error
+}
+
+// VersionType represents the version and type information for container runtime
+type VersionType struct {
+  // The version of container runtime
+  Version string
+  // The type of container runtime
+  Type    string
+}
+
+// PodMeta represents
+type PodMeta struct {
+  // The uid of the pod
+  UID string
+  // The name of the pod
+  Name string
+  // The namespace of the pod
+  Namespace string
+}
+```
+
+**Kubelet network plugin interface**
+
+```go
+// NetworkPlugin is an interface to network plugins for the kubelet
+type NetworkPlugin interface {
+  // Init initializes the plugin.  This will be called exactly once
+  // before any other methods are called.
+  Init(host Host, containerRuntime string) error
+
+  // Name returns the plugin's name. This will be used when searching
+  // for a plugin by name, e.g.
+  Name() string
+
+  // SetUpPod is the method called when the pod is launched.
+  SetUpPod(namespace, name, podID string) error
+
+  // TearDownPod is called before a pod be deleted
+  TearDownPod(namespace, name, podID string) error
+
+  // Status is the method called to obtain the ipv4 or ipv6 addresses of the pod
+  Status(namespace, name, podID string) (*PodNetworkStatus, error)
+}
+```
+
+## Implementation
+
+The implementation will be based on [grpc](http://www.grpc.io).
```

Comment:
grpc is prefered because of its performance and multi-language support. Shell command maybe not acceptable because there are too much data to pass in and out.

I think `remote` has the same meaning with `external`.